File: Task3C/furniture.cpp
#include "furniture.h"

// Default constructor
Furniture::Furniture() {
    this->material = "";
    this->dimensions = "";
}

// Parameterised constructor
Furniture::Furniture(string name, int price, Supplier* supplier, int volume, int productID, string material, string dimensions): Product(name, price, supplier, volume, productID) {
    this->material = material;
    this->dimensions = dimensions;
}

// Getters and setters
string Furniture::getMaterial() {
    return this->material;
}

string Furniture::getDimensions() {
    return this->dimensions;
}

void Furniture::setMaterial(string material) {
    this->material = material;
}

void Furniture::setDimensions(string dimensions) {
    this->dimensions = dimensions;
}

// Display product information
void Furniture::display_product_info() {
    Product::display_product_info();
    cout << "Material: " << this->material << endl;
    cout << "Dimensions: " << this->dimensions << endl;
}

// Save product information to file
void Furniture::save(ostream& out) const {
    Product::save(out);
    out << material << "," << dimensions << "\n";
}

// Load product information from file
void Furniture::load(istream& in) {
    Product::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    material = token;

    getline(ss, token, ',');
    dimensions = token;
}

File: Task2A_B/task2a_b/src/main/java/com/cps2004/Furniture.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public abstract class Furniture extends Product{
    private String material;
    private String dimensions;

    // Default constructor
    public Furniture() {
        super();
        material = "";
        dimensions = "";
    }

    // Parameterized constructor
    public Furniture(String name, int id, Supplier supplier, int price, int volume, String material, String dimensions) {
        super(name, id, supplier, price, volume);
        this.material = material;
        this.dimensions = dimensions;
    }

    // Getters and setters
    public String getMaterial() { return material; }
    public String getDimensions() { return dimensions; }
    public void setMaterial(String material) { this.material = material; }
    public void setDimensions(String dimensions) { this.dimensions = dimensions; }

    @Override
    // Calculate discounted price
    public abstract double calculate_discount(String month, int quantity);

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product material: " + material);
        System.out.println("Product dimensions: " + dimensions);
    }

    // Save product information to file
    public void save(ObjectNode newItem) {
        super.save(newItem);
        newItem.put("material", this.material);
        newItem.put("dimensions", this.dimensions);
    }

    // Load product information from file
    public void load(JsonNode node) {
        super.load(node);
        material = node.get("material").asText();
        dimensions = node.get("dimensions").asText();
    }
}
